File: 卡码/代码随想录/Java/src/com/kurumi/codeThoughts/动态规划/_02_背包问题/_02_完全背包/LeetCode_0279.java
package com.kurumi.codeThoughts.动态规划._02_背包问题._02_完全背包;

import java.util.Arrays;

/**
 * @Author kurumi
 * @Date 2026-07-03 上午11:52:20
 * @Description 279. 完全平方数
 */
public class LeetCode_0279 {
    static class Solution {
        public int numSquares(int n) {
            int[] dp = new int[n + 1];
            Arrays.fill(dp, Integer.MAX_VALUE - 1);
            dp[0] = 0;
            for (int i = 1; i * i <= n; i++) {
                for (int j = i * i; j <= n; j++) {
                    dp[j] = Math.min(dp[j], dp[j - i * i] + 1);
                }
            }
            return dp[n];
        }
    }
}

File: 卡码/代码随想录/Go/动态规划/02-背包问题/02-完全背包/LeetCode_0279.go
// 279. 完全平方数
package main

import (
	"math"
)

func numSquares(n int) int {
	dp := make([]int, n+1)
	for i := 0; i < n+1; i++ {
		dp[i] = math.MaxInt32 - 1
	}
	dp[0] = 0
	for i := 1; i*i <= n; i++ {
		for j := i * i; j <= n; j++ {
			dp[j] = int(math.Min(float64(dp[j]), float64(dp[j-i*i]+1)))
		}
	}
	return dp[n]
}
